File: backend/app/sync.py
"""The sync core: idempotent intake and incremental push.

The correctness of the whole offline-first design rests on this file, so the logic is deliberately plain.

Invariants
------
1. However often an op_id is replayed, its side effect happens once.
2. "Record the sync_op" and "produce the side effect" both happen or
   neither does -- they have to share a transaction. A crash in between
   would leave a "recorded but never applied" hole that replay then skips
   as a duplicate, and the row is gone for good.
3. One failing op must not take the batch down -- SAVEPOINT isolates it.
"""

import json
import logging

from sqlalchemy import text
from sqlalchemy.orm import Session

from .schemas import SyncOpIn
from .services.checks import (
    add_order_lines,
    close_check,
    merge_checks,
    modify_check,
    open_check,
    open_togo_check,
    restore_check,
    add_payment,
    set_payment,
    transfer_check,
    void_check,
    void_order_line,
)
from .services.buffet import record_tray_event

log = logging.getLogger(__name__)

# The roles each entity needs. Authorisation on the write path reads **only**
# this table, never anything the client claims about itself.
_FRONT = frozenset({"front_employee", "front_manager", "admin"})
# Editing and voiding **make money disappear** -- managers and the owner only
_FRONT_MANAGER = frozenset({"front_manager", "admin"})

_HANDLERS: dict[str, frozenset[str]] = {
    "open_check": _FRONT,
    "close_check": _FRONT,
    # Transferring, merging and changing the payment method move no money, so
    # regular staff can do them -- needing a manager every time is enough
    # friction that people stop recording anything
    "transfer_check": _FRONT,
    "merge_checks": _FRONT,
    "set_payment": _FRONT,
    # Top-ups are open to regular staff too: money in hand has to be recorded
    # now, and waiting for a manager at peak means it never gets recorded
    "add_payment": _FRONT,
    # To-go and adding dishes are daily work
    "open_togo_check": _FRONT,
    "add_order_lines": _FRONT,
    "void_order_line": _FRONT,
    "modify_check": _FRONT_MANAGER,
    "void_check": _FRONT_MANAGER,
    "restore_check": _FRONT_MANAGER,
    # Refills: **the front and the kitchen can both record them**. A cook
    # tapping while refilling is the natural case, but the person who notices
    # an empty tray is usually a server -- kitchen-only would lose most of the
    # "ran empty" events, and those are the right-hand end of the censoring interval.
    "tray_event": _FRONT | {"kitchen"},
}


def _apply_effect(db: Session, op: SyncOpIn, user) -> None:
    """Produce the business side effect. Has to share the transaction with the sync_op insert."""
    if op.entity == "open_check":
        open_check(db, op.op_id, op.payload, op.client_ts, user.id if user else None)

    elif op.entity == "close_check":
        close_check(db, op.payload, op.client_ts)

    elif op.entity == "modify_check":
        modify_check(db, op.payload, op.client_ts)

    elif op.entity == "void_check":
        void_check(db, op.payload, op.client_ts, user.id if user else None)

    elif op.entity == "restore_check":
        restore_check(db, op.payload, op.client_ts, user.id if user else None)

    elif op.entity == "transfer_check":
        transfer_check(db, op.payload, op.client_ts)

    elif op.entity == "merge_checks":
        merge_checks(db, op.payload, op.client_ts)

    elif op.entity == "set_payment":
        set_payment(db, op.payload, op.client_ts)

    elif op.entity == "add_payment":
        add_payment(db, op.payload, op.client_ts)

    elif op.entity == "open_togo_check":
        open_togo_check(db, op.op_id, op.payload, op.client_ts,
                        user.id if user else None)

    elif op.entity == "add_order_lines":
        add_order_lines(db, op.payload, op.client_ts)

    elif op.entity == "void_order_line":
        void_order_line(db, op.payload, op.client_ts)

    elif op.entity == "tray_event":
        record_tray_event(db, op.payload, op.client_ts, user.id if user else None)

    else:
        raise ValueError(f"Unknown entity: {op.entity}")


def apply_ops(db: Session, client_id: str, ops: list[SyncOpIn], user=None):
    """Apply ops one by one, idempotently. Returns (applied, duplicate, rejected).

    `user` comes from the access token the server verified, **not from anything
    the client says about itself**. Every op records its user_id -- walkouts,
    comps and voids have to be traceable to a person or there is no control at all.
    """
    applied: list = []
    duplicate: list = []
    rejected: list = []

    for op in ops:
        allowed = _HANDLERS.get(op.entity)
        if allowed is None:
            rejected.append({"op_id": op.op_id, "reason": f"Unknown entity: {op.entity}"})
            continue

        # ⚠️ Authorisation happens here, not in the front end.
        # Ops queued on a client while offline carry no role claim --
        # the role comes only from the access token the server verified.
        if user is not None and user.role not in allowed:
            rejected.append(
                {
                    "op_id": op.op_id,
                    "reason": f"{user.role} may not perform {op.entity}",
                }
            )
            continue

        try:
            # SAVEPOINT: if this one blows up it rolls back alone, and the ops before it stand
            with db.begin_nested():
                # The heart of idempotency: on a key conflict do nothing and
                # **return no row**. Got a row = we are the first writer = the
                # side effect is ours. No row = someone (or an earlier replay) already did it.
                row = db.execute(
                    text(
                        """
                        INSERT INTO sync_op
                            (op_id, client_id, user_id, entity, op_type, payload,
                             client_seq, client_ts)
                        VALUES
                            (:op_id, :client_id, :user_id, :entity, :op_type,
                             CAST(:payload AS jsonb), :client_seq, :client_ts)
                        ON CONFLICT (op_id) DO NOTHING
                        RETURNING op_id
                        """
                    ),
                    {
                        "op_id": str(op.op_id),
                        "client_id": client_id,
                        "user_id": user.id if user else None,
                        "entity": op.entity,
                        "op_type": op.op_type,
                        "payload": json.dumps(op.payload),
                        "client_seq": op.client_seq,
                        "client_ts": op.client_ts,
                    },
                ).first()

                if row is None:
                    duplicate.append(op.op_id)
                    continue

                _apply_effect(db, op, user)

                db.execute(
                    text("UPDATE sync_op SET applied_at = now() WHERE op_id = :op_id"),
                    {"op_id": str(op.op_id)},
                )
                applied.append(op.op_id)

        except Exception as exc:  # noqa: BLE001 -- one failure must not stop the batch
            log.warning("op %s rejected: %s", op.op_id, exc)
            rejected.append({"op_id": op.op_id, "reason": f"{type(exc).__name__}: {exc}"})

    db.commit()
    return applied, duplicate, rejected


def log_truncated(db: Session, since_cursor: int) -> bool:
    """The client consumed up to since_cursor, but **not one record at or below**
    **that seq is left in the log**.

    In normal operation that cannot happen: a cursor of N means 1..N existed,
    and one of them is still there. Only a log that was deleted wholesale (test
    data cleared, retention archiving) produces it.

    An increment is then not enough -- the client's mirror holds checks the
    server no longer has, and since sync is **append-only** nothing later will
    ever remove them. The device would show them forever. So it starts over.

    ⚠️ The test cannot be `since_cursor > MAX(seq)`: seq is a bigserial and
       DELETE does not wind it back. After a purge, one write from another
       device is seq 80, so the device sitting at 79 looks fine and never agrees again.
    """
    if since_cursor <= 0:
        return False
    hit = db.execute(
        text("SELECT 1 FROM sync_op WHERE seq <= :n LIMIT 1"), {"n": since_cursor}
    ).first()
    return hit is None


def fetch_changes(
    db: Session,
    since_cursor: int,
    client_id: str,
    limit: int = 500,
    resync: bool = False,
):
    """Fetch applied changes after since_cursor that came from **other devices**.

    Filtering out its own keeps a client from applying what it just wrote.

    With `resync=True` **nothing is filtered**: the client just emptied its
    mirror, so its own checks have to be sent again or a chunk is simply missing.
    """
    rows = db.execute(
        text(
            """
            SELECT s.seq, s.op_id, s.client_id, s.entity, s.client_ts, s.payload,
                   u.display_name AS user_display
              FROM sync_op s
              LEFT JOIN app_user u ON u.id = s.user_id
             WHERE s.seq > :since
               AND s.applied_at IS NOT NULL
               AND (:resync OR s.client_id <> :client_id)
             ORDER BY s.seq
             LIMIT :limit
            """
        ),
        {
            "since": since_cursor,
            "client_id": client_id,
            "limit": limit,
            "resync": resync,
        },
    ).mappings().all()

    # The cursor only advances to the last row actually returned -- using
    # MAX(seq) would skip changes in the middle whenever LIMIT truncated.
    next_cursor = rows[-1]["seq"] if rows else since_cursor

    # But an untruncated page means everything up to max(seq) is consumed
    # (whatever was filtered out came from this device), so it is safe to jump.
    # Otherwise every op this device wrote gets rescanned on every later sync.
    if len(rows) < limit:
        max_applied = db.execute(
            text("SELECT COALESCE(MAX(seq), 0) FROM sync_op WHERE applied_at IS NOT NULL")
        ).scalar_one()
        next_cursor = max(next_cursor, max_applied)

    return [dict(r) for r in rows], next_cursor

File: frontend/src/sync.ts
import { authFetch } from './auth'
import { applyCheckOp } from './checks'
import { applyTrayOp } from './trays'
import db, {
  clientId,
  getMeta,
  nextClientSeq,
  setMeta,
  uuid,
  type OutboxOp,
} from './db'

/** How many ops one sync carries at most; matches SyncRequest.ops max_length on the server. */
const BATCH = 200

type RemoteChange = {
  op_id: string
  entity: string
  client_ts: string
  /** Display name the server attaches -- a device only knows that for its own ops */
  user_display?: string | null
  payload: Record<string, unknown>
}

export type SyncResult = {
  applied: number
  duplicate: number
  rejected: { op_id: string; reason: string }[]
  changes: number
  cursor: number
}

/**
 * The two kinds of sync failure are nothing alike and must never be conflated:
 *
 * - `offline`: no network. **This is a normal state**; the data is safe in the outbox.
 * - `auth`: the session expired and someone has to sign in again. Still queued safely.
 * - `error`: connected, but the server failed. This is the one that needs a person.
 *
 * Showing all three as "sync failed" was a real problem -- a server who reads
 * "failed" redoes the work, and in a restaurant a lying status is worse than a crash.
 */
export type SyncFailure = { kind: 'offline' | 'auth' | 'error'; message: string }

/** Hard timeout for one request. Fetch has **no default timeout** -- see STUCK_MS below. */
const FETCH_TIMEOUT = 10_000

/**
 * The in-flight watchdog.
 *
 * What this cost once (six records lost on a real device):
 * airplane mode goes on with a fetch in flight. Fetch has no default timeout,
 * so that promise **never settles** -- `inFlight` stays set forever and every
 * later sync() hands back the same dead promise, **never issuing a request
 * again**. The timer keeps firing, Caddy never hears anything, and the UI sits
 * on whatever it last reported.
 *
 * Two defences: (1) AbortController on the fetch, (2) this watchdog behind it.
 */
const STUCK_MS = 30_000

/** Only one sync may run at a time, or the same batch of ops goes out twice. */
let inFlight: Promise<SyncResult | null> | null = null
let inFlightAt = 0

/**
 * The write path: **write locally, then queue**.
 *
 * Both tables are written in one Dexie transaction -- either the mirror and the
 * outbox both have it, or neither does. Otherwise you get a ghost record the UI
 * shows and sync never sends.
 */
export async function enqueue(
  entity: string,
  payload: Record<string, unknown>,
  /** Callers may bring their own op_id -- a check uses it as its identity */
  opId?: string,
): Promise<string> {
  const op_id = opId ?? uuid()
  const client_ts = new Date().toISOString()

  await db.transaction('rw', db.outbox, db.trays, db.meta, async () => {
    const seq = await nextClientSeq()

    const op: OutboxOp = {
      op_id,
      entity,
      op_type: 'insert',
      client_seq: seq,
      client_ts,
      payload,
    }
    await db.outbox.add(op)

    // A refill's local mirror row and its outbox entry are written in **one
    // transaction**. Split, they produce "the record went out but the board does
    // not show it" -- and "last refilled N minutes ago" is the only feedback there is, so people tap again.
    if (entity === 'tray_event') {
      const back = Number(payload.minutes_ago ?? 0) || 0
      await db.trays.add({
        op_id,
        dish_id: Number(payload.dish_id),
        kind: String(payload.event_type) as 'refill' | 'half' | 'empty',
        at: new Date(Date.parse(client_ts) - back * 60_000).toISOString(),
        synced: 0,
        remote: 0,
      })
    }
  })

  // Send it as soon as possible: the less time in the outbox, the less a broken device loses
  void sync()
  return op_id
}

export async function sync(): Promise<SyncResult | null> {
  // Only reuse one that is **really still running**. Past the watchdog window it
  // counts as dead: drop it and start again -- resending is safe, op_id is idempotent.
  if (inFlight && Date.now() - inFlightAt < STUCK_MS) return inFlight

  if (inFlight) console.warn('[sync] the previous sync looks stuck; dropping it and retrying')

  inFlightAt = Date.now()
  const p = doSync().finally(() => {
    // Only clear it while still the current one, or a later sync gets cleared by an earlier one
    if (inFlight === p) inFlight = null
  })
  inFlight = p
  return p
}

/**
 * Take the ops to send.
 *
 * ⚠️ **Never use `db.outbox.orderBy('client_seq')`.**
 * That is an index query, and IndexedDB **silently excludes** records whose
 * index key is invalid (NaN / undefined / null) -- while `count()` still counts
 * them. The result: the UI says "4 pending" and the request body is `ops: []`,
 * with those four stuck forever and no error anywhere. That is how they were lost on a real device.
 *
 * Reading everything and sorting in JS **depends on no index, so nothing can hide.**
 */
async function takePending(): Promise<OutboxOp[]> {
  const all = await db.outbox.toArray()

  // Repair bad historical rows on the way past: give an invalid client_seq a
  // legal value, or they stay invisible on the next pass even if this one sends them
  const broken = all.filter((o) => !Number.isFinite(o.client_seq))
  if (broken.length) {
    console.warn(`[sync] repairing ${broken.length} records with an invalid client_seq`)
    let base = await nextClientSeq()
    await db.transaction('rw', db.outbox, db.meta, async () => {
      for (const o of broken) {
        o.client_seq = base++
        await db.outbox.put(o)
      }
      await setMeta('client_seq', base)
    })
  }

  return all
    .sort((a, b) => (a.client_seq ?? 0) - (b.client_seq ?? 0))
    .slice(0, BATCH)
}

/**
 * The server says its log no longer holds anything before our cursor (`reset`).
 *
 * Sync is **append-only**: what the server deleted never arrives as a change,
 * so those checks stay in the local mirror and the screen shows checks that no
 * longer exist. The only right move is to drop the mirror, zero the cursor and pull again.
 *
 * ⚠️ **Only synced=1 rows go.** synced=0 means it is still in the outbox and
 *    has never been sent -- a check the store just entered that the server has
 *    never seen, so clearing it really would lose data.
 *    The outbox and the dead letter queue are not touched at all.
 */
async function resetMirror(): Promise<void> {
  await db.transaction('rw', db.checks, db.events, db.trays, db.meta, async () => {
    await db.checks.filter((c) => c.synced === 1).delete()
    await db.events.filter((e) => e.synced === 1).delete()
    await db.trays.filter((t) => t.synced === 1).delete()
    await setMeta('cursor', 0)
    // Stored rather than kept in memory: reloading halfway through a re-pull
    // still has to start over, or the mirror misses the part the server "only sends to others".
    await setMeta('resync', 1)
  })
  console.warn('[sync] the server log was truncated; rebuilding the local mirror')
}

async function doSync(): Promise<SyncResult | null> {
  const ops = await takePending()
  const since = await getMeta<number>('cursor', 0)
  const cid = await clientId()
  const resync = (await getMeta<number>('resync', 0)) === 1

  // Send even with an empty outbox: this is the only moment other devices' changes get pulled
  const body = JSON.stringify({
    client_id: cid,
    since_cursor: since,
    ops,
    resync,
  })

  let res: Response
  const ctrl = new AbortController()
  const timer = window.setTimeout(() => ctrl.abort(), FETCH_TIMEOUT)
  try {
    res = await authFetch('/api/sync', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body,
      signal: ctrl.signal,
    })
  } catch (e) {
    // A request aborted on timeout **may well have arrived and been processed**,
    // with only the response lost. That is fine -- the replay comes back as a duplicate, which is what op_id is for.
    const aborted = (e as Error)?.name === 'AbortError'
    const f: SyncFailure = {
      kind: 'offline',
      message: aborted ? 'Request timed out, queued for retry' : 'Offline, queued',
    }
    throw f
  } finally {
    window.clearTimeout(timer)
  }

  if (res.status === 401) {
    // authFetch already tried to refresh, so a 401 here means the session is really gone.
    // **Never clear the outbox** -- unsynced records belong to the store and
    // still have to go out after signing in again.
    const f: SyncFailure = { kind: 'auth', message: 'Session expired, please sign in again' }
    throw f
  }

  if (!res.ok) {
    // Do not clear the outbox -- retry next time. Resending is safe, op_id is idempotent.
    //
    // ⚠️ A reachable proxy with a dead upstream returns 502/503/504, so fetch
    //    **succeeds** and the catch above never runs. To staff that is no
    //    different from being offline: the data is queued and resends itself,
    //    and nobody has to do anything.
    //    A red "error" makes them think it broke and re-enter the check.
    //    408/429 (timeout / rate limit) are the same -- they pass on their own.
    const transient = [408, 425, 429, 500, 502, 503, 504].includes(res.status)
    const f: SyncFailure = transient
      ? { kind: 'offline', message: `Server unavailable (${res.status}), queued` }
      : { kind: 'error', message: `Server HTTP ${res.status}` }
    throw f
  }

  const data = await res.json()

  // applied and duplicate both mean "the server has it", so both leave the outbox.
  // A duplicate is not an error; it is the replay mechanism working.
  const settled: string[] = [...data.applied, ...data.duplicate]

  // Explicitly rejected by the server: **it must not sit in the outbox retrying**
  // forever, or the outbox never empties, "pending" never reaches zero and every
  // round wastes a send.
  // Move it to the dead letter queue for a person to handle.
  const rejected = (data.rejected ?? []) as { op_id: string; reason: string }[]
  let remoteChanges: RemoteChange[] = []


  // Dexie only takes an array scope past five tables
  await db.transaction(
    'rw',
    [db.outbox, db.events, db.meta, db.deadletter, db.checks, db.trays],
    async () => {
      if (rejected.length) {
        const byId = new Map(ops.map((o) => [o.op_id, o]))
        for (const r of rejected) {
          const op = byId.get(r.op_id)
          await db.deadletter.put({
            op_id: r.op_id,
            entity: op?.entity ?? '?',
            payload: op?.payload ?? {},
            reason: r.reason,
            failed_at: new Date().toISOString(),
          })
        }
        await db.outbox.bulkDelete(rejected.map((r) => r.op_id))
      }

      if (settled.length) {
        await db.outbox.bulkDelete(settled)
        await Promise.all(settled.map((id) => db.events.update(id, { synced: 1 })))
        // A check's check_uuid is the op_id of the op that created it
        await Promise.all(settled.map((id) => db.checks.update(id, { synced: 1 })))
        await Promise.all(settled.map((id) => db.trays.update(id, { synced: 1 })))
      }

      // Other devices' changes. On a reset this batch is empty by definition,
      // and the mirror is about to be cleared -- so nothing is applied.
      remoteChanges = data.reset ? [] : (data.changes as RemoteChange[])

      if (!data.reset) {
        // The cursor advances last: a crash in between just repeats a round, and repeating beats skipping
        await setMeta('cursor', data.cursor)

        // A full re-pull is only done at an **empty response**. A full batch means
        // more is coming (the server sends 500 at most), and clearing resync now
        // would make the next page filter out this device's own ops again -- leaving a gap.
        if (resync && data.changes.length === 0) await setMeta('resync', 0)
      }
    },
  )

  if (data.reset) {
    // Order matters: the transaction above already settled this batch out of the
    // outbox and marked the matching local checks synced=1. Marking has to come
    // first -- otherwise they stay in the mirror as synced=0 with no op left in
    // the outbox, which is a ghost check that can never sync.
    await resetMirror()
    // Immediately go again and pull back what the server actually has.
    // Not waiting for the next heartbeat (up to 20 seconds) -- the screen is empty until then.
    setTimeout(() => void sync(), 0)
    return {
      applied: data.applied.length,
      duplicate: data.duplicate.length,
      rejected: data.rejected,
      changes: 0,
      cursor: 0,
    }
  }

  // Check changes are applied outside the transaction -- applyCheckOp reads the
  // catalog (also in the meta table), and overlapping table scopes would deadlock
  for (const c of remoteChanges) {
    if (c.entity === 'tray_event') {
      await applyTrayOp(c.op_id, c.payload, c.client_ts, {
        synced: 1,
        remote: 1,
        who: c.user_display ?? undefined,
      })
    } else if (c.entity === 'ping_event') {
      await db.events.put({
        op_id: c.op_id,
        label: String(c.payload.label ?? ''),
        // Use the source device's client_ts, not "when it arrived" --
        // otherwise a backlog of offline records all sorts to the top and the timeline is wrong
        created_at: c.client_ts,
        synced: 1,
        remote: 1,
      })
    } else {
      await applyCheckOp(c.entity, c.op_id, c.payload, c.client_ts, {
        synced: 1,
        remote: 1,
        who: c.user_display ?? undefined,
      })
    }
  }

  return {
    applied: data.applied.length,
    duplicate: data.duplicate.length,
    rejected: data.rejected,
    changes: data.changes.length,
    cursor: data.cursor,
  }
}

/** Retry interval while the outbox has a backlog -- fast, because staff are watching it clear */
const POLL_PENDING = 4_000
/** Idle heartbeat -- only to pull other devices' changes, so it can be slow */
const POLL_IDLE = 20_000

/**
 * When it fires.
 *
 * ⚠️ iOS has no Background Sync -- nothing ever replays in the background.
 * So it has to sync the moment the app **comes back to the foreground**; that is the critical hook on an iPad.
 */
export function installSyncTriggers(
  onResult?: (r: SyncResult | null, failure?: SyncFailure) => void,
) {
  let timer: number | undefined
  let stopped = false

  const schedule = async () => {
    if (stopped) return
    const pending = await db.outbox.count()
    window.clearTimeout(timer)
    timer = window.setTimeout(run, pending > 0 ? POLL_PENDING : POLL_IDLE)
  }

  const run = () => {
    sync()
      .then((r) => onResult?.(r))
      .catch((e: unknown) => {
        const f: SyncFailure =
          e && typeof e === 'object' && 'kind' in e
            ? (e as SyncFailure)
            : { kind: 'error', message: String(e) }
        onResult?.(null, f)
      })
      .finally(schedule)
  }

  window.addEventListener('online', run)
  // ⚠️ iOS has no Background Sync -- returning to the foreground is the critical hook
  document.addEventListener('visibilitychange', () => {
    if (document.visibilityState === 'visible') run()
  })
  run()

  return () => {
    stopped = true
    window.removeEventListener('online', run)
    window.clearTimeout(timer)
  }
}
